drink/src/runtime/pallet_contracts_debugging.rs

```diff
@@ -0,0 +1,114 @@
+//! This module provides all the necessary elements for supporting contract debugging directly in
+//! the contracts pallet.
+//!
+//! # Smart-contract developer <-> pallet-contracts interaction flow
+//!
+//! The interaction between end-user and runtime is as follows:
+//! 1. At some points during execution, the pallet invokes some callback through its configuration
+//! parameter `Debug`.
+//! 2. In order to forward the callback outside the runtime, `Debug` will call a runtime interface,
+//! that will then forward the call further to the proper runtime extension.
+//! 3. The runtime extension can be fully controlled by the end-user. It just has to be registered
+//! in the runtime.
+//!
+//! So, in brief: pallet-contracts -> runtime interface -> runtime extension
+//!              |<-----------runtime side-------------->|<---user side--->|
+//!
+//! # Passing objects between runtime and runtime extension
+//!
+//! Unfortunately, runtime interface that lies between runtime and the end-user accepts only
+//! very simlpe argument types and those that implement some specific traits. This means that
+//! usually, complext objects will be passed in their encoded form (`Vec<u8>` obtained with scale
+//! encoding).
+
+use pallet_contracts::debug::{CallSpan, ExportedFunction, Tracing};
+use pallet_contracts_primitives::ExecReturnValue;
+use sp_externalities::{decl_extension, ExternalitiesExt};
+use sp_runtime_interface::runtime_interface;
+
+use crate::runtime::Runtime;
+
+type AccountIdOf<R> = <R as frame_system::Config>::AccountId;
+
+/// The trait that allows injecting custom logic to handle contract debugging directly in the
+/// contracts pallet.
+pub trait DebugExtT {
+    /// Called after a contract call is made.
+    fn after_call(
+        &self,
+        _contract_address: Vec<u8>,
+        _is_call: bool,
+        _input_data: Vec<u8>,
+        _result: Vec<u8>,
+    ) {
+    }
```

Comment:
due to API change here https://github.com/paritytech/substrate/pull/14789 it was easier to skip `before_call` for now; I will add it as a follow up (to avoid increasing complexity of this PR)